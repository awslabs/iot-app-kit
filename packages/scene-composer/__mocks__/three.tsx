/**
 * Mock implementation of R3F (React Three Fiber) components with the string name for a clean snapshot test.
 */
module.exports = {
  ...jest.requireActual('three'),
  audioListener: 'audioListener',
  positionalAudio: 'positionalAudio',

  mesh: 'mesh',
  instancedMesh: 'instancedMesh',
  scene: 'scene',
  sprite: 'sprite',
  lOD: 'lOD',
  skinnedMesh: 'skinnedMesh',
  skeleton: 'skeleton',
  bone: 'bone',
  lineSegments: 'lineSegments',
  lineLoop: 'lineLoop',
  points: 'points',
  group: 'group',
  immediateRenderObject: 'immediateRenderObject',

  // cameras
  camera: 'camera',
  perspectiveCamera: 'perspectiveCamera',
  orthographicCamera: 'orthographicCamera',
  cubeCamera: 'cubeCamera',
  arrayCamera: 'arrayCamera',

  // geometry
  instancedBufferGeometry: 'instancedBufferGeometry',
  bufferGeometry: 'bufferGeometry',
  boxBufferGeometry: 'boxBufferGeometry',
  circleBufferGeometry: 'circleBufferGeometry',
  coneBufferGeometry: 'coneBufferGeometry',
  cylinderBufferGeometry: 'cylinderBufferGeometry',
  dodecahedronBufferGeometry: 'dodecahedronBufferGeometry',
  extrudeBufferGeometry: 'extrudeBufferGeometry',
  icosahedronBufferGeometry: 'icosahedronBufferGeometry',
  latheBufferGeometry: 'latheBufferGeometry',
  octahedronBufferGeometry: 'octahedronBufferGeometry',
  parametricBufferGeometry: 'parametricBufferGeometry',
  planeBufferGeometry: 'planeBufferGeometry',
  polyhedronBufferGeometry: 'polyhedronBufferGeometry',
  ringBufferGeometry: 'ringBufferGeometry',
  shapeBufferGeometry: 'shapeBufferGeometry',
  sphereBufferGeometry: 'sphereBufferGeometry',
  tetrahedronBufferGeometry: 'tetrahedronBufferGeometry',
  textBufferGeometry: 'textBufferGeometry',
  torusBufferGeometry: 'torusBufferGeometry',
  torusKnotBufferGeometry: 'torusKnotBufferGeometry',
  tubeBufferGeometry: 'tubeBufferGeometry',
  wireframeGeometry: 'wireframeGeometry',
  parametricGeometry: 'parametricGeometry',
  tetrahedronGeometry: 'tetrahedronGeometry',
  octahedronGeometry: 'octahedronGeometry',
  icosahedronGeometry: 'icosahedronGeometry',
  dodecahedronGeometry: 'dodecahedronGeometry',
  polyhedronGeometry: 'polyhedronGeometry',
  tubeGeometry: 'tubeGeometry',
  torusKnotGeometry: 'torusKnotGeometry',
  torusGeometry: 'torusGeometry',
  textGeometry: 'textGeometry',
  sphereGeometry: 'sphereGeometry',
  ringGeometry: 'ringGeometry',
  planeGeometry: 'planeGeometry',
  latheGeometry: 'latheGeometry',
  shapeGeometry: 'shapeGeometry',
  extrudeGeometry: 'extrudeGeometry',
  edgesGeometry: 'edgesGeometry',
  coneGeometry: 'coneGeometry',
  cylinderGeometry: 'cylinderGeometry',
  circleGeometry: 'circleGeometry',
  boxGeometry: 'boxGeometry',

  // materials
  material: 'material',
  shadowMaterial: 'shadowMaterial',
  spriteMaterial: 'spriteMaterial',
  rawShaderMaterial: 'rawShaderMaterial',
  shaderMaterial: 'shaderMaterial',
  pointsMaterial: 'pointsMaterial',
  meshPhysicalMaterial: 'meshPhysicalMaterial',
  meshStandardMaterial: 'meshStandardMaterial',
  meshPhongMaterial: 'meshPhongMaterial',
  meshToonMaterial: 'meshToonMaterial',
  meshNormalMaterial: 'meshNormalMaterial',
  meshLambertMaterial: 'meshLambertMaterial',
  meshDepthMaterial: 'meshDepthMaterial',
  meshDistanceMaterial: 'meshDistanceMaterial',
  meshBasicMaterial: 'meshBasicMaterial',
  meshMatcapMaterial: 'meshMatcapMaterial',
  lineDashedMaterial: 'lineDashedMaterial',
  lineBasicMaterial: 'lineBasicMaterial',

  // primitive
  primitive: 'primitive',

  // lights and other
  light: 'light',
  spotLightShadow: 'spotLightShadow',
  spotLight: 'spotLight',
  pointLight: 'pointLight',
  rectAreaLight: 'rectAreaLight',
  hemisphereLight: 'hemisphereLight',
  directionalLightShadow: 'directionalLightShadow',
  directionalLight: 'directionalLight',
  ambientLight: 'ambientLight',
  lightShadow: 'lightShadow',
  ambientLightProbe: 'ambientLightProbe',
  hemisphereLightProbe: 'hemisphereLightProbe',
  lightProbe: 'lightProbe',

  // helpers
  spotLightHelper: 'spotLightHelper',
  skeletonHelper: 'skeletonHelper',
  pointLightHelper: 'pointLightHelper',
  hemisphereLightHelper: 'hemisphereLightHelper',
  gridHelper: 'gridHelper',
  polarGridHelper: 'polarGridHelper',
  directionalLightHelper: 'directionalLightHelper',
  cameraHelper: 'cameraHelper',
  boxHelper: 'boxHelper',
  box3Helper: 'box3Helper',
  planeHelper: 'planeHelper',
  arrowHelper: 'arrowHelper',
  axesHelper: 'axesHelper',

  // textures
  texture: 'texture',
  videoTexture: 'videoTexture',
  dataTexture: 'dataTexture',
  dataTexture3D: 'dataTexture3D',
  compressedTexture: 'compressedTexture',
  cubeTexture: 'cubeTexture',
  canvasTexture: 'canvasTexture',
  depthTexture: 'depthTexture',

  // misc
  raycaster: 'raycaster',
  vector2: 'vector2',
  vector3: 'vector3',
  vector4: 'vector4',
  euler: 'euler',
  matrix3: 'matrix3',
  matrix4: 'matrix4',
  quaternion: 'quaternion',
  bufferAttribute: 'bufferAttribute',
  instancedBufferAttribute: 'instancedBufferAttribute',
  color: 'color',
  fog: 'fog',
  fogExp2: 'fogExp2',
  shape: 'shape',
};
