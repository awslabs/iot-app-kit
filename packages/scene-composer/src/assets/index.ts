export * from './anchors';
export * from './viewpoints';
