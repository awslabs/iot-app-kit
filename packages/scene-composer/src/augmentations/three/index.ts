export * from './Anchor';
export * from './Viewpoint';
