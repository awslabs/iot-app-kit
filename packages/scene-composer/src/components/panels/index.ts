export { SceneRulesPanel } from './SceneRulesPanel';
export { SceneHierarchyPanel } from './SceneHierarchyPanel';
export { SceneNodeInspectorPanel } from './SceneNodeInspectorPanel';
export { SettingsPanel } from './SettingsPanel';
export { TopBar } from './TopBar';
