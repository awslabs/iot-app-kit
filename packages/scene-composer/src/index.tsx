export * from './interfaces';

export * from './common/errors';

export * from './components/SceneComposerInternal';

export * from './SceneViewer';

export type { Scene } from './models/SceneModels';

// Augmentation related interfaces
export * from './augmentations/three';

export * from './common/GlobalSettings';

export * from './utils/dataFrameLabelUtils';

export * from './utils/dataBindingTemplateUtils';

export * from './common/constants';
