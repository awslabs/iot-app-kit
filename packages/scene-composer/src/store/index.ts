export * from './StoreOperations';
export * from './Store';
