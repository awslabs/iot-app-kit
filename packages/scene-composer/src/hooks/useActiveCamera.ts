import { useCallback } from 'react';
import * as THREE from 'three';

import { useSceneComposerId } from '../common/sceneComposerIdContext';
import { useEditorState } from '../store';
import { CameraControlMode, CameraSettings } from '../interfaces';
import { DEFAULT_CAMERA_OPTIONS, DEFAULT_CAMERA_POSITION, DEFAULT_CAMERA_TARGET } from '../common/constants';
import { CameraType, Transform } from '../models/SceneModels';

const useActiveCamera = () => {
  const sceneComposerId = useSceneComposerId();
  const {
    activeCameraSettings,
    setActiveCameraSettings,
    setCameraTarget,
    isViewing,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  } = useEditorState(sceneComposerId);

  const defaultCameraSettings: CameraSettings = {
    cameraType: CameraType.Perspective,
    fov: DEFAULT_CAMERA_OPTIONS.fov,
    far: DEFAULT_CAMERA_OPTIONS.far,
    near: DEFAULT_CAMERA_OPTIONS.near,
    zoom: 1,
    transform: {
      position: DEFAULT_CAMERA_POSITION,
    },
  };

  const calculateTargetOrGetDefault = (transform: Transform) => {
    if (transform.rotation) {
      const obj3D = new THREE.Object3D();
      obj3D.position.set(...transform.position);
      obj3D.rotation.set(...transform.rotation);
      obj3D.updateMatrix();
      obj3D.updateMatrixWorld(true);

      // Place an object to look at in front of the parent and get it's world position
      const targetObj3D = new THREE.Object3D();
      obj3D.add(targetObj3D);
      targetObj3D.position.set(0, 0, -10);
      targetObj3D.updateMatrix();
      targetObj3D.updateMatrixWorld(true);
      const target = targetObj3D.getWorldPosition(new THREE.Vector3());
      return target.toArray();
    }

    return DEFAULT_CAMERA_TARGET;
  };

  const setActiveCamera = useCallback(
    (cameraSettings?: CameraSettings, mode: CameraControlMode = 'transition', forceSetTarget = false) => {
      const settings = cameraSettings ?? defaultCameraSettings;
      setActiveCameraSettings(settings);

      const cameraTarget = {
        position: settings.transform.position,
        target: calculateTargetOrGetDefault(settings.transform),
      };

      if (forceSetTarget || isViewing()) {
        setCameraTarget(cameraTarget, mode);
      }
    },
    [setActiveCameraSettings, setCameraTarget],
  );

  return {
    activeCameraSettings,
    setActiveCameraSettings: setActiveCamera,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  };
};

export default useActiveCamera;
