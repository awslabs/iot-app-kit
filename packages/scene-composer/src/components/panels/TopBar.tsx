import React, { FC, useCallback, useContext, useMemo } from 'react';
import styled from 'styled-components';
import { ButtonDropdown, SpaceBetween } from '@awsui/components-react';
import { useIntl } from 'react-intl';

import { KnownComponentType } from '../../interfaces';
import { sceneComposerIdContext } from '../../common/sceneComposerIdContext';
import { ICameraComponentInternal, useStore } from '../../store';
import useActiveCamera from '../../hooks/useActiveCamera';
import { findComponentByType } from '../../utils/nodeUtils';
import { getCameraSettings } from '../../utils/cameraUtils';

// TODO: SpaceBetween is not intended to have custom styles, we should refactor this to use our own spacing component
// if these styles are really needed.
const StyledSpaceBetween = styled(SpaceBetween)`
  flex: 1;
  justify-content: right;
`;

export const TopBar: FC = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const nodeMap = useStore(sceneComposerId)((state) => state.document.nodeMap);
  const getSceneNodeByRef = useStore(sceneComposerId)((state) => state.getSceneNodeByRef);
  const getObject3DBySceneNodeRef = useStore(sceneComposerId)((state) => state.getObject3DBySceneNodeRef);
  const { setActiveCameraSettings } = useActiveCamera();
  const intl = useIntl();

  const cameraItems = useMemo(() => {
    return Object.values(nodeMap)
      .filter((node) => {
        return findComponentByType(node, KnownComponentType.Camera);
      })
      .map((cameraNode) => {
        return {
          id: cameraNode.ref,
          text: cameraNode!.name,
        };
      });
  }, [nodeMap]);

  const hasCameraView = cameraItems.length > 0;
  const showTopBar = hasCameraView;

  const setActiveCameraOnItemClick = useCallback(
    ({ detail }) => {
      const node = getSceneNodeByRef(detail.id);
      const cameraComponent = findComponentByType(node, KnownComponentType.Camera) as ICameraComponentInternal;
      const object3D = getObject3DBySceneNodeRef(detail.id);

      setActiveCameraSettings(getCameraSettings(object3D, cameraComponent), 'transition', true);
    },
    [setActiveCameraSettings],
  );

  if (showTopBar) {
    return (
      <StyledSpaceBetween direction='horizontal' size='xxs'>
        {hasCameraView && (
          <ButtonDropdown data-testid='camera-views' items={cameraItems} onItemClick={setActiveCameraOnItemClick}>
            {intl.formatMessage({ defaultMessage: 'Camera View', description: 'camera views dropdown button text' })}
          </ButtonDropdown>
        )}
      </StyledSpaceBetween>
    );
  }

  return <></>;
};
