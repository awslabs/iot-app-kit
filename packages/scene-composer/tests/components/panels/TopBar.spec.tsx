import React from 'react';
import { render } from '@testing-library/react';
import wrapper from '@awsui/components-react/test-utils/dom';
import * as THREE from 'three';

import { TopBar } from '../../../src/components/panels/TopBar';
import { useStore } from '../../../src/store';
import {
  COMPOSER_FEATURES,
  DEFAULT_CAMERA_OPTIONS,
  DEFAULT_CAMERA_POSITION,
  KnownComponentType,
  setFeatureConfig,
} from '../../../src';
import useActiveCamera from '../../../src/hooks/useActiveCamera';

jest.mock('@awsui/components-react', () => ({
  ...jest.requireActual('@awsui/components-react'),
}));

jest.mock('../../../src/hooks/useActiveCamera', () =>
  jest.fn().mockReturnValue({ setActiveCameraSettings: jest.fn() }),
);

const toggleMotionIndicatorVisibilityMock = jest.fn();
const cameraSettings = {
  cameraType: 'Perspective',
  fov: DEFAULT_CAMERA_OPTIONS.fov,
  far: DEFAULT_CAMERA_OPTIONS.far,
  near: DEFAULT_CAMERA_OPTIONS.near,
  zoom: 1,
};
const cameraComponent = {
  ...cameraSettings,
  type: KnownComponentType.Camera,
};

const cameraNode = {
  ref: 'camera-uuid',
  name: 'Camera',
  transform: {
    position: DEFAULT_CAMERA_POSITION,
  },
  components: [cameraComponent],
};

const node = {
  ref: 'test-uuid',
  name: 'Model',
  transform: {
    position: [1, 1, 1],
  },
  components: [{ type: KnownComponentType.ModelRef }],
};

const object3D = new THREE.Object3D();
object3D.position.set(5, 5, 5);
object3D.rotation.set(0, 0, 0);
object3D.scale.set(1, 1, 1);

const getObject3DBySceneNodeRef = jest.fn().mockReturnValue(object3D);
const getComponentRefByType = jest.fn();

const baseState = {
  noHistoryStates: {
    motionIndicatorVisible: true,
    toggleMotionIndicatorVisibility: toggleMotionIndicatorVisibilityMock,
  },
  getSceneNodeByRef: jest.fn().mockReturnValue(cameraNode),
  getObject3DBySceneNodeRef,
  getComponentRefByType,
  document: {
    nodeMap: {
      'test-uuid': node,
    },
  },
} as any;

describe('<TopBar />', () => {
  it('should render with motion indicator off option', () => {
    getComponentRefByType.mockReturnValue({ type: KnownComponentType.MotionIndicator });
    useStore('default').setState({
      ...baseState,
      noHistoryStates: {
        motionIndicatorVisible: false,
      },
    } as any);

    const { container } = render(<TopBar />);
    const polarisWrapper = wrapper(container);

    const dropDown = polarisWrapper.findButtonDropdown('[data-testid="view-options"]');
    dropDown?.openDropdown();
    const item = dropDown?.findItemById(KnownComponentType.MotionIndicator);

    // has no icon
    expect(item?.find('svg')).toBeNull();
  });

  it('should not render without motion indicator or camera view', () => {
    getComponentRefByType.mockReturnValue({});
    useStore('default').setState(baseState);

    const { container } = render(<TopBar />);
    expect(container).toMatchSnapshot();
  });

  it('should render View Options Drop down with motion indicator', () => {
    getComponentRefByType.mockReturnValue({ type: KnownComponentType.MotionIndicator });
    useStore('default').setState({ ...baseState });

    const { container } = render(<TopBar />);
    expect(container).toMatchSnapshot();
  });

  it('should render Cameras Drop down with camera view', () => {
    getComponentRefByType.mockReturnValue({});
    const document = {
      nodeMap: {
        'camera-uuid': cameraNode,
      },
    };

    useStore('default').setState({ ...baseState, document });

    const { container } = render(<TopBar />);
    expect(container).toMatchSnapshot();
  });

  it('should render both drop downs with motion indicator and camera view', () => {
    getComponentRefByType.mockReturnValue({ type: KnownComponentType.MotionIndicator });
    const document = {
      nodeMap: {
        'camera-uuid': cameraNode,
      },
    };

    useStore('default').setState({ ...baseState, document });

    const { container } = render(<TopBar />);
    expect(container).toMatchSnapshot();
  });

  it('should toggle motion indicator visibility', () => {
    getComponentRefByType.mockReturnValue({ type: KnownComponentType.MotionIndicator });
    useStore('default').setState(baseState);

    const { container } = render(<TopBar />);
    const polarisWrapper = wrapper(container);

    const dropDown = polarisWrapper.findButtonDropdown('[data-testid="view-options"]');
    dropDown?.openDropdown();
    let item = dropDown?.findItemById(KnownComponentType.MotionIndicator);

    // has checked icon
    expect(item?.find('[data-testid="CheckedIcon"]')).toBeDefined();

    item?.click();
    dropDown?.openDropdown();
    item = dropDown?.findItemById(KnownComponentType.MotionIndicator);

    expect(item?.find('svg')).toBeTruthy();
  });

  it('should call setActiveCameraSettings when camera clicked', () => {
    getComponentRefByType.mockReturnValue({});
    const document = {
      nodeMap: {
        'camera-uuid': cameraNode,
      },
    };

    useStore('default').setState({ ...baseState, document });
    setFeatureConfig({ [COMPOSER_FEATURES.CameraView]: true });
    const { setActiveCameraSettings } = useActiveCamera();

    const { container } = render(<TopBar />);
    const polarisWrapper = wrapper(container);

    const dropDown = polarisWrapper.findButtonDropdown('[data-testid="camera-views"]');
    dropDown?.openDropdown();
    const item = dropDown?.findItemById('camera-uuid');

    item?.click();

    expect(setActiveCameraSettings).toBeCalledWith(
      {
        ...cameraSettings,
        transform: {
          position: object3D.getWorldPosition(new THREE.Vector3()).toArray(),
          rotation: new THREE.Euler()
            .setFromQuaternion(object3D.getWorldQuaternion(new THREE.Quaternion()))
            .toVector3()
            .toArray(),
          scale: object3D.getWorldScale(new THREE.Vector3()).toArray(),
        },
      },
      'transition',
      true,
    );
  });
});
