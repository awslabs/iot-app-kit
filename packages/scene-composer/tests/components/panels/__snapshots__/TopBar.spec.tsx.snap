// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> should not render without motion indicator or camera view 1`] = `<div />`;

exports[`<TopBar /> should render Cameras Drop down with camera view 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: right;
  -webkit-justify-content: right;
  -ms-flex-pack: right;
  justify-content: right;
}

<div>
  <div
    class="c0 awsui_root_18582_66aol_93 awsui_horizontal_18582_66aol_108 awsui_horizontal-xxs_18582_66aol_116"
  >
    <div
      class="awsui_child_18582_66aol_97 awsui_child-horizontal-xxs_18582_66aol_151"
    >
      <div
        class="awsui_button-dropdown_sne0l_14t3j_93"
        data-testid="camera-views"
      >
        <div
          class="awsui_root_qwoo0_1jaxg_143"
        >
          <div
            class=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="awsui_button_vjswe_10957_101 awsui_variant-normal_vjswe_10957_122"
              type="button"
            >
              <span
                class="awsui_content_vjswe_10957_97"
              >
                Cameras
              </span>
              <span
                class="awsui_icon_vjswe_10957_901 awsui_icon-right_vjswe_10957_906 awsui_rotate-down_sne0l_14t3j_129 awsui_icon_h11ix_31bp4_98 awsui_size-normal-mapped-height_h11ix_31bp4_151 awsui_size-normal_h11ix_31bp4_147 awsui_variant-normal_h11ix_31bp4_219"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="filled stroke-linejoin-round"
                    d="M4 5h8l-4 6-4-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            tabindex="-1"
          />
          <div>
            <div
              tabindex="-1"
            />
            <div
              class="awsui_dropdown_qwoo0_1jaxg_93 awsui_with-limited-width_qwoo0_1jaxg_215"
              data-animating="false"
              data-open="false"
            >
              <div
                class="awsui_dropdown-content-wrapper_qwoo0_1jaxg_93"
              >
                <div
                  class="awsui_ie11-wrapper_qwoo0_1jaxg_257"
                >
                  <div
                    class="awsui_dropdown-content_qwoo0_1jaxg_93"
                  >
                    <ul
                      class="awsui_options-list_19gcf_1wdy7_93 awsui_decrease-top-margin_19gcf_1wdy7_113"
                      role="menu"
                      style="position: static;"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TopBar /> should render View Options Drop down with motion indicator 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: right;
  -webkit-justify-content: right;
  -ms-flex-pack: right;
  justify-content: right;
}

<div>
  <div
    class="c0 awsui_root_18582_66aol_93 awsui_horizontal_18582_66aol_108 awsui_horizontal-xxs_18582_66aol_116"
  >
    <div
      class="awsui_child_18582_66aol_97 awsui_child-horizontal-xxs_18582_66aol_151"
    >
      <div
        class="awsui_button-dropdown_sne0l_14t3j_93"
        data-testid="view-options"
      >
        <div
          class="awsui_root_qwoo0_1jaxg_143"
        >
          <div
            class=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="awsui_button_vjswe_10957_101 awsui_variant-normal_vjswe_10957_122"
              type="button"
            >
              <span
                class="awsui_content_vjswe_10957_97"
              >
                View Options
              </span>
              <span
                class="awsui_icon_vjswe_10957_901 awsui_icon-right_vjswe_10957_906 awsui_rotate-down_sne0l_14t3j_129 awsui_icon_h11ix_31bp4_98 awsui_size-normal-mapped-height_h11ix_31bp4_151 awsui_size-normal_h11ix_31bp4_147 awsui_variant-normal_h11ix_31bp4_219"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="filled stroke-linejoin-round"
                    d="M4 5h8l-4 6-4-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            tabindex="-1"
          />
          <div>
            <div
              tabindex="-1"
            />
            <div
              class="awsui_dropdown_qwoo0_1jaxg_93 awsui_with-limited-width_qwoo0_1jaxg_215"
              data-animating="false"
              data-open="false"
            >
              <div
                class="awsui_dropdown-content-wrapper_qwoo0_1jaxg_93"
              >
                <div
                  class="awsui_ie11-wrapper_qwoo0_1jaxg_257"
                >
                  <div
                    class="awsui_dropdown-content_qwoo0_1jaxg_93"
                  >
                    <ul
                      class="awsui_options-list_19gcf_1wdy7_93 awsui_decrease-top-margin_19gcf_1wdy7_113"
                      role="menu"
                      style="position: static;"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TopBar /> should render both drop downs with motion indicator and camera view 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: right;
  -webkit-justify-content: right;
  -ms-flex-pack: right;
  justify-content: right;
}

<div>
  <div
    class="c0 awsui_root_18582_66aol_93 awsui_horizontal_18582_66aol_108 awsui_horizontal-xxs_18582_66aol_116"
  >
    <div
      class="awsui_child_18582_66aol_97 awsui_child-horizontal-xxs_18582_66aol_151"
    >
      <div
        class="awsui_button-dropdown_sne0l_14t3j_93"
        data-testid="view-options"
      >
        <div
          class="awsui_root_qwoo0_1jaxg_143"
        >
          <div
            class=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="awsui_button_vjswe_10957_101 awsui_variant-normal_vjswe_10957_122"
              type="button"
            >
              <span
                class="awsui_content_vjswe_10957_97"
              >
                View Options
              </span>
              <span
                class="awsui_icon_vjswe_10957_901 awsui_icon-right_vjswe_10957_906 awsui_rotate-down_sne0l_14t3j_129 awsui_icon_h11ix_31bp4_98 awsui_size-normal-mapped-height_h11ix_31bp4_151 awsui_size-normal_h11ix_31bp4_147 awsui_variant-normal_h11ix_31bp4_219"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="filled stroke-linejoin-round"
                    d="M4 5h8l-4 6-4-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            tabindex="-1"
          />
          <div>
            <div
              tabindex="-1"
            />
            <div
              class="awsui_dropdown_qwoo0_1jaxg_93 awsui_with-limited-width_qwoo0_1jaxg_215"
              data-animating="false"
              data-open="false"
            >
              <div
                class="awsui_dropdown-content-wrapper_qwoo0_1jaxg_93"
              >
                <div
                  class="awsui_ie11-wrapper_qwoo0_1jaxg_257"
                >
                  <div
                    class="awsui_dropdown-content_qwoo0_1jaxg_93"
                  >
                    <ul
                      class="awsui_options-list_19gcf_1wdy7_93 awsui_decrease-top-margin_19gcf_1wdy7_113"
                      role="menu"
                      style="position: static;"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="awsui_child_18582_66aol_97 awsui_child-horizontal-xxs_18582_66aol_151"
    >
      <div
        class="awsui_button-dropdown_sne0l_14t3j_93"
        data-testid="camera-views"
      >
        <div
          class="awsui_root_qwoo0_1jaxg_143"
        >
          <div
            class=""
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="awsui_button_vjswe_10957_101 awsui_variant-normal_vjswe_10957_122"
              type="button"
            >
              <span
                class="awsui_content_vjswe_10957_97"
              >
                Cameras
              </span>
              <span
                class="awsui_icon_vjswe_10957_901 awsui_icon-right_vjswe_10957_906 awsui_rotate-down_sne0l_14t3j_129 awsui_icon_h11ix_31bp4_98 awsui_size-normal-mapped-height_h11ix_31bp4_151 awsui_size-normal_h11ix_31bp4_147 awsui_variant-normal_h11ix_31bp4_219"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="filled stroke-linejoin-round"
                    d="M4 5h8l-4 6-4-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            tabindex="-1"
          />
          <div>
            <div
              tabindex="-1"
            />
            <div
              class="awsui_dropdown_qwoo0_1jaxg_93 awsui_with-limited-width_qwoo0_1jaxg_215"
              data-animating="false"
              data-open="false"
            >
              <div
                class="awsui_dropdown-content-wrapper_qwoo0_1jaxg_93"
              >
                <div
                  class="awsui_ie11-wrapper_qwoo0_1jaxg_257"
                >
                  <div
                    class="awsui_dropdown-content_qwoo0_1jaxg_93"
                  >
                    <ul
                      class="awsui_options-list_19gcf_1wdy7_93 awsui_decrease-top-margin_19gcf_1wdy7_113"
                      role="menu"
                      style="position: static;"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
